### PR TITLE
User Authentication, Authorization, and API with Spotify

### PR DESCRIPTION
**Pull request by aaronli03**
## What
- users can now authorize our application to connect with their Spotify accounts
- our authorization method implements PKCE
- api call to retrieve necessary user data from Spotify endpoint
- api call to retrieve song recommendations from Spotify endpoint
- api call to retrieve search results from Spotify endpoint

## Why
- a user needs to be connected to make calls to Spotify endpoint
- PKCE offers additional security
- display name is needed to personalize user experience
- song recommendations used to display contents to user
- search results used to allow for users to customize filter options

## Impact
- Will customer experience be significantly altered? Yes
- Does the code change cover more than one task? No

## Testing
- tested locally by running `npm run dev`

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -8,7 +8,8 @@
 		"resolveJsonModule": true,
 		"skipLibCheck": true,
 		"sourceMap": true,
-		"strict": true
+		"strict": true,
+		"ignoreDeprecations": "5.0"
```
Comment (byjackli) [inline tsconfig.json:12]: Why is `ignoreDeprecations` set?

File: src/api/auth.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+export function isLoggedIn(): boolean {
+	return userExist;
+}
+
+export function logout(): void {
+	userExist = false;
+	localStorage.clear();
+	window.location.href = `http://127.0.0.1:5173/`;
```
Comment (byjackli) [inline src/api/auth.ts:76]: Why is the location hard-coded?

File: src/api/auth.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+let userExist = false;
+const clientId = process.env.CLIENT_ID;
+const clientSecret = process.env.CLIENT_SECRET;
+const redirectUri = process.env.REDIRECT_URI;
+
+const scope = `user-read-private user-read-email playlist-read-private playlist-read-collaborative playlist-modify-public playlist-modify-private`;
+
+function generateRandomString(length: number) {
```
Comment (byjackli) [inline src/api/auth.ts:16]: where are you getting all these functions from? and where are their respective tests?

**Review by byjackli — COMMENTED** [review]
- [ ] split into smaller commits
- [ ] make sure there are good tests for each function

File: src/api/auth.ts (hunk 3 above)
Comment (aaronli03) [reply src/api/auth.ts:16]: this function is used to generate a state to be passed in the auth flow
the others are used to implement pkce

File: src/api/auth.ts (hunk 2 above)
Comment (aaronli03) [reply src/api/auth.ts:76]: it will be replaced later on by using firebase cloud functions

File: tsconfig.json (hunk 1 above)
Comment (aaronli03) [reply tsconfig.json:12]: it is a temporary fix for the error below:
"Option 'importsNotUsedAsValues' is deprecated and will stop functioning in TypeScript 5.5. Specify compilerOption '"ignoreDeprecations": "5.0"' to silence this error.
  Use 'verbatimModuleSyntax' instead.ts"

File: src/api/auth.ts (hunk 3 above)
Comment (byjackli) [reply src/api/auth.ts:16]: Did you come up with these functions yourself? There's no official package to pull from?

File: tsconfig.json (hunk 1 above)
Comment (byjackli) [reply tsconfig.json:12]: What's the long-term fix?

Comment (aaronli03) [reply tsconfig.json:12]: 'importsNotUsedAsValues' set to true. I will update this

File: src/api/auth.ts (hunk 3 above)
Comment (aaronli03) [reply src/api/auth.ts:16]: No I followed instructions but I will update it to use the npm package "pkce-challenge"

Comment (aaronli03) [reply src/api/auth.ts:16]: Actually I cannot find one that works

**Pull request closed without merging**